### PR TITLE
[Docking] Feature: support for not needing to set dock instances

### PR DESCRIPTION
**Pull request by jcarlosgm30**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #4415 |
| Primary OS tested on | Ubuntu 22.04 |
| Robotic platform tested on | Gazebo Simulation |
| Does this PR contain AI generated software? | No |

---

## Description of contribution in a few bullet points
The changes made are as follows:
* [Support for not needing to set `dock_instances`](https://github.com/ros-navigation/navigation2/commit/15b085167ae4ee7682c636c7987a66a9b94488e1)
* [Improve some log messages in `dock_database.cpp`](https://github.com/ros-navigation/navigation2/commit/24337173b228c22a963be3ae5b98760adf1e6f18)
* [Comment the dock instances out from `nav2_params`](https://github.com/ros-navigation/navigation2/commit/ec9c012e8b4dd1e62d8a0aeaac3b106f3280a984)

These changes enable the system to work in the following use cases:
* Setting the dock instances in the parameters or database yaml file.
* Not setting the dock instances in these files but using the dock pose in the action request or the reload database service.

Also, the reload database service is available in any case where at least one dock plugin has been defined.

## Description of documentation updates required from your changes
* The documentation may need to be updated to reflect the "Not setting ..." use case.

---

## Future work that may be required in bullet points
* The unit tests may need to be updated to address new changes

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_bringup/params/nav2_params.yaml [hunk 1]
```diff
@@ -407,11 +407,12 @@ docking_server:
       filter_coef: 0.1
 
     # Dock instances
-    docks: ['home_dock']  # Input your docks here
-    home_dock:
-      type: 'simple_charging_dock'
-      frame: map
-      pose: [0.0, 0.0, 0.0]
+    # The following example illustrates how you should configure the docks instances.
```
Comment (SteveMacenski) [inline nav2_bringup/params/nav2_params.yaml:410]:
```suggestion
    # The following example illustrates configuring dock instances.
```

File: nav2_docking/opennav_docking/src/dock_database.cpp [hunk 2]
```diff
@@ -71,12 +82,20 @@ void DockDatabase::reloadDbCb(
...
   if (utils::parseDockFile(request->filepath, node_.lock(), dock_instances)) {
     dock_instances_ = dock_instances;
     response->success = true;
+    RCLCPP_INFO(
+      node->get_logger(),
+      "Dock database reloaded from file %s.", request->filepath.c_str());
     return;
   }
+
+  RCLCPP_ERROR(
```
Comment (SteveMacenski) [inline nav2_docking/opennav_docking/src/dock_database.cpp:96]: Remove. Failure messages are already logged by `parseDockFile`

File: nav2_docking/opennav_docking/src/dock_database.cpp [hunk 3]
```diff
@@ -71,12 +82,20 @@ void DockDatabase::reloadDbCb(
   const std::shared_ptr<nav2_msgs::srv::ReloadDockDatabase::Request> request,
   std::shared_ptr<nav2_msgs::srv::ReloadDockDatabase::Response> response)
 {
+  auto node = node_.lock();
   DockMap dock_instances;
   if (utils::parseDockFile(request->filepath, node_.lock(), dock_instances)) {
```
Comment (SteveMacenski) [inline nav2_docking/opennav_docking/src/dock_database.cpp:87]: Update to use this new `node`

File: nav2_docking/opennav_docking/src/dock_database.cpp [hunk 4]
```diff
@@ -99,6 +118,17 @@ Dock * DockDatabase::findDock(const std::string & dock_id)
 
 Dock * DockDatabase::findDockInstance(const std::string & dock_id)
 {
+  if (dock_instances_.empty()) {
```
Comment (SteveMacenski) [inline nav2_docking/opennav_docking/src/dock_database.cpp:121]: This is not necessary, please remove. If the database is empty, `find` results in the map's `end` which returns nullptr already.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Generally looks good! Just some small cleanup!

File: nav2_docking/opennav_docking/src/dock_database.cpp (hunk 4 above)
Comment (jcarlosgm30) [reply nav2_docking/opennav_docking/src/dock_database.cpp:121]: I only add it to be more verbose for the user

File: nav2_docking/opennav_docking/src/dock_database.cpp (hunk 3 above)
Comment (jcarlosgm30) [reply nav2_docking/opennav_docking/src/dock_database.cpp]: done

File: nav2_bringup/params/nav2_params.yaml (hunk 1 above)
Comment (jcarlosgm30) [reply nav2_bringup/params/nav2_params.yaml]: done

File: nav2_docking/opennav_docking/src/dock_database.cpp (hunk 2 above)
Comment (jcarlosgm30) [reply nav2_docking/opennav_docking/src/dock_database.cpp]: done

File: nav2_docking/opennav_docking/src/dock_database.cpp (hunk 4 above)
Comment (jcarlosgm30) [reply nav2_docking/opennav_docking/src/dock_database.cpp]: done

**Review by SteveMacenski — APPROVED** [review]

**Comment by jcarlosgm30** [comment]
Yes, that test fails because there is no dock plugin initialized so the `/reload_database` service will not be created.

The old logic was that the `/reload_database` service was only created when `getDockPlugins` and `getDockInstances` returned false. As we discussed, the new logic creates the `/reload_database` service whenever at least a dock plugin is initialized. 

So I think it will be enough to initialize a dock plugin in the test.

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Fabulous, thank you!

**Pull request merged**